lldb/source/Target/Process.cpp

```diff
@@ -2550,6 +2550,21 @@ ModuleSP Process::ReadModuleFromMemory(const FileSpec &file_spec,
   }
   ModuleSP module_sp(new Module(file_spec, ArchSpec()));
   if (module_sp) {
+    if (size_to_read == 0) {
+      // Default to 8192 in case we can't find a memory region.
+      size_to_read = 0x2000;
+      MemoryRegionInfo range_info;
+      Status error(GetMemoryRegionInfo(header_addr, range_info));
+      if (error.Success()) {
+        // We found a memory region, set the range of bytes ro read to read to
+        // the end of the memory region. This should be enough to contain the
+        // file header and important bits.
+        const auto &range = range_info.GetRange();
```

Comment:
We don't care about the code, we care about getting all of the program headers. Without this change, my test doesn't work and the test verifies that this code is working. Prior to this, this was defaulting to 512 bytes, so we would sometimes get data back for the program headers if they started within the first 512 bytes, and sometimes we would get part of the program headers and they would be cut short because we didn't have all of the data. Other times we would get no bytes for the program headers. By getting the first memory segment for a binary, this fixed all 3 cases as it at least ensures we have enough info for the program headers. Once we have all program headers, we can now end up reading anything else we need from process memory, so any data pointed to by the program headers can be fetched from the process memory. I can remove this change and do the extra work in the ObjectFileELF.cpp when we load from memory, but I figured this would be useful for other things like mach-o and COFF too.

So in your example above, check to see if all program headers are in the first LOAD segment, if that is true, then this patch will work just fine on those.

Reply:
They are, but that's not my point. My point is that this first segment can be really large (like one gigabyte or even more), and if that's the case, we probably don't want to read all of it.

Doing the extra read in ObjectFileELF sounds good to me, as there we can look at the elf header, and load exactly only the program header bytes we care about.